optuna/_gp/_gp.py

```diff
@@ -0,0 +1,191 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+import math
+import typing
+from typing import Callable
+
+import numpy as np
+import scipy.optimize
+import torch
+
+
+# This GP implementation uses the following notation:
+# X[len(trials), len(params)]: observed parameter values.
+# Y[len(trials)]: observed objective values.
+# x[(batch_len,) len(params)]: parameter value to evaluate. Possibly batched.
+# cov_fX_fX[len(trials), len(trials)]: kernel matrix of X = V[f(X)]
+# cov_fx_fX[(batch_len,) len(trials)]: kernel matrix of x and X = Cov[f(x), f(X)]
+# cov_fx_fx: kernel value (scalar) of x = V[f(x)].
+#     Since we use a Matern 5/2 kernel, we assume this value to be a constant.
+# cov_Y_Y_inv[len(trials), len(trials)]: inv of the covariance matrix of Y = (V[f(X) + noise])^-1
+# cov_Y_Y_inv_Y[len(trials)]: cov_Y_Y_inv @ Y
+# max_Y: maximum of Y (Note that we transform the objective values such that it is maximized.)
+# d2: squared distance between two points
+
+
+class Matern52Kernel(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx: typing.Any, squared_distance: torch.Tensor) -> torch.Tensor:  # type: ignore
+        sqrt5d = torch.sqrt(5 * squared_distance)
+        exp_part = torch.exp(-sqrt5d)
+        val = exp_part * ((5 / 3) * squared_distance + sqrt5d + 1)
+        # Notice that the derivative is taken w.r.t. d^2, but not w.r.t. d.
+        deriv = (-5 / 6) * (sqrt5d + 1) * exp_part
+        ctx.save_for_backward(deriv)
+        return val
+
+    @staticmethod
+    def backward(ctx: typing.Any, grad: torch.Tensor) -> torch.Tensor:  # type: ignore
+        # Let x be squared_distance, f(x) be forward(ctx, x), and g(f) be a provided function,
+        # then deriv := df/dx, grad := dg/df, and deriv * grad = df/dx * dg/df = dg/dx.
+        (deriv,) = ctx.saved_tensors
+        return deriv * grad
+
+
+# This is the value of the Matern 5/2 kernel at squared_distance=0.
+MATERN_KERNEL0 = 1.0
+
+
+def matern52_kernel_from_squared_distance(squared_distance: torch.Tensor) -> torch.Tensor:
+    # sqrt5d = sqrt(5 * squared_distance)
+    # exp(sqrt5d) * (1/3 * sqrt5d ** 2 + sqrt5d + 1)
+    #
+    # We cannot let PyTorch differentiate the above expression because
+    # the gradient runs into 0/0 at squared_distance=0.
+    return Matern52Kernel.apply(squared_distance)  # type: ignore
+
+
+@dataclass(frozen=True)
+class KernelParams:
```

Comment:
Can we add a method like this and use it instead of `MATERN_KERNEL0`?

```python
    @classmethod
    def kernel_value_for_zero_distance(cls) -> float:
        return 1.0
```

Reply:
Sorry, please ignore it!